cabby/data/metagraph/utils.py

```diff
@@ -169,75 +144,81 @@ def convert_multidi_to_weighted_undir_graph(
-
-  # Add geometry data to OSM nodes.
-  osm_geometries = {}
+
+  # Step 2: Add all geometries to the graph.
   for _, row in osm_map.nodes.iterrows():
-    assert metagraph.has_node(row["index"]), "Map.node index %s not in graph" % row["index"]
-    osm_geometries[row["index"]] = row["geometry"]
-  nx.set_node_attributes(metagraph, osm_geometries, name="geometry")
-
-  # Set useful node attributes.
-  name_to_osmid = {osmid: name for name, osmid in osmid_to_name.items()}
-  nx.set_node_attributes(metagraph, name_to_osmid, name="osmid")
-  nx.set_node_attributes(metagraph, name_to_point, name="geometry")
-  nx.set_node_attributes(metagraph, name_to_wikidata, name="wikidata")
-
-  # Add conceptual edges from wd_relations.
-  wikidata_to_name = {wikidata: name for
-                      name, wikidata in name_to_wikidata.items()}
-  for _, row in wd_relations.iterrows():
-    place_node_name = wikidata_to_name[row["place"]]
-    concept_node_name = "%s_%s" % (row["instanceLabel"], row["instance"])
-    metagraph.add_edge(place_node_name, concept_node_name, weight=1.0)
+    metagraph.nodes[row["osmid"]]["geometry"] = row["geometry"]
+  for _, row in osm_map.poi.iterrows():
+    if "geometry" in metagraph.nodes[row["osmid"]]:
+      continue
+    assert isinstance(row["geometry"], Point)
+    metagraph.nodes[row["osmid"]]["geometry"] = row["geometry"]
 
-  # S2 cell nodes and edges.
+  # Step 3: Add important POI attributes.
+  attributes_to_add = collections.defaultdict(dict)
+  wikidata_to_nodeid = {}  # Needed for wikidata concept node additions.
+  for _, row in osm_map.poi.iterrows():
+    node_id = row["osmid"]
+    for field in IMPORTANT_POI_METADATA_FIELDS:
+      if isinstance(row[field], str):
+        attributes_to_add[node_id][field] = row[field]
+        if field == "wikidata":
+          wikidata_to_nodeid[row[field]] = node_id
+  nx.set_node_attributes(metagraph, values=attributes_to_add)
+
+  # Step 4: Add conceptual nodes, attributes, and edges to the graph.
+  attributes_to_add = collections.defaultdict(dict)
+  for _, row in wd_relations.iterrows():
+    # Add edge.
+    place_node_id = wikidata_to_nodeid[row["place"]]
+    concept_node_id = row["instance"]
+    metagraph.add_edge(node_id, concept_node_id, weight=1.0)
+    # Add attributes.
+    attributes_to_add[place_node_id]["name"] = row["placeLabel"]
+    attributes_to_add[concept_node_id]["name"] = row["instanceLabel"]
+    attributes_to_add[concept_node_id]["wikidata"] = row["place"]
+  nx.set_node_attributes(metagraph, values=attributes_to_add)
+
+  # Step 5: Add S2 nodes and edges
   edges_to_add = []
   for node, data in metagraph.nodes.data():
-    if not data or "geometry" not in data:
+    if "geometry" not in data:
       continue
     geometry = data["geometry"]
     for level in s2_node_levels:
       s2_cell_id = util.cellid_from_point(geometry, level)
-      edges_to_add.append((node, s2_cell_id))
+      s2_cell_node_id = "S2_L%d_%s" % (level, s2_cell_id)
```

Comment:
If I'm not mistaken, the level can be determined from the s2 cell id so including Lx in the node id is likely redundant. (That said, it might still make sense to keep it for readability.)

Reply:
Left a TODO to replace the prefix with a S2 cell type attribute in e2087a6